src/event-log.c

```diff
@@ -0,0 +1,398 @@
+ *
+ * @param log_level log level, either from GLogLevelFlags, or a user-defined level
+ * @param fields key–value pairs of structured data forming the log message.
+ *               [array length=n_fields]
+ * @param n_fields number of elements in the fields array
+ * @param verbose whether or not to print known log fields, too.
+ *
+ * @return newly-allocated formatted string (without trailing newline)
+ */
+static gchar *event_log_format_fields_readable(GLogLevelFlags log_level,
+		const GLogField *fields, gsize n_fields, gboolean verbose)
+{
+	g_autoptr(GDateTime) now = NULL;
+	g_autofree gchar *now_formatted = NULL;
+	GString *gstring;
+
+	gstring = g_string_new(NULL);
+	now = g_date_time_new_now_utc();
+	now_formatted = g_date_time_format(now, "%Y-%m-%dT%H:%M:%SZ");
+	g_string_append_printf(gstring, "%s: ", now_formatted);
+
+	for (gsize j = 0; j < n_fields; j++) {
+		const GLogField *ifield = &fields[j];
+		const gchar *message = NULL;
+
+		/* print message */
+		if (g_strcmp0(ifield->key, "MESSAGE") == 0) {
+			message = ifield->value;
+			g_string_append_printf(gstring, "%s", message);
+			break;
+		}
+	}
+
+	/* skip other fields if non-verbose */
+	if (!verbose)
+		return g_string_free(gstring, FALSE);
+
+	/* collect verbose fields in order of appearance */
+	for (gsize j = 0; j < n_fields; j++) {
+		const GLogField *ifield = &fields[j];
+		const gchar *key = NULL;
+		const gchar *value = NULL;
+
+		if (g_strcmp0(ifield->key, "TRANSACTION_ID") == 0) {
+			key = "transaction ID";
+			value = ifield->value;
+		} else if (g_strcmp0(ifield->key, "BUNDLE_HASH") == 0) {
+			key = "bundle hash";
+			value = ifield->value;
+		} else if (g_strcmp0(ifield->key, "BOOT_ID") == 0) {
+			key = "boot ID";
+			value = ifield->value;
+		}
+
+		if (key) {
+			g_string_append_printf(gstring, "\n                      %s: %s", key, value);
+		}
+	}
+
+	return g_string_free(gstring, FALSE);
+}
+
+/**
+ * Rotates log files.
+ *
+ * If a maximum size is configured and the loggers current size + provided
+ * 'next_len' exceed this maximum size, the method will close the current
+ * output stream, and copy the file to a .1 copy.
+ *
+ * All previously existing rotation files will be rotated by +1.
+ *
+ * If max-files is set, rotation is limited to max-files files.
+ *
+ * @param logger Logger to rotate file for
+ * @param next_len Length of next data to add.
+ *        This is used to determine if the log file needs to be rotated.
+ */
+static void rotate_logfiles(REventLogger* logger, goffset next_len)
+{
+	g_autoptr(GFile) logfile = NULL;
+	g_autofree gchar *rotfile = NULL;
+	g_autoptr(GError) ierror = NULL;
+
+	g_return_if_fail(logger);
+
+	/* method is no-op if no maxsize is configured */
+	if (!logger->maxsize)
+		return;
+
+	/* return if size limit not reached yet */
+	if (logger->filesize + next_len <= logger->maxsize)
+		return;
+
+	/* Close open stream */
+	g_clear_pointer(&logger->logstream, g_object_unref);
```

Comment:
Should we flush before closing? Or even `fsync()`?

Reply:
From the Glib `GOutputStream` documentation:

> Closing the stream will implicitly cause a flush.

Since a `GFileOutputStream` is also a `GOutputStream`, this will apply here, too.

However, calling `fsync()` should be possible when assuming having a `GUnixOutputStream` as we do for the update handlers already.